rts/motoko-rts/src/stabilization/reader_writer.rs

```diff
@@ -0,0 +1,287 @@
+//! Buffered read/write access on stable memory.
+//! Supporting Cheney's to-space in stable memory.
+
+use core::{
+    cmp::min,
+    mem::{size_of, zeroed},
+};
+
+use crate::{
+    mem_utils::memcpy_bytes,
+    stable_mem::{ic0_stable64_read, ic0_stable64_write, PAGE_SIZE},
+    types::Bytes,
+};
+
+use super::grant_stable_space;
+
+const FREE_PAGE_INDEX: u64 = u64::MAX;
+
+struct StableMemoryPage {
+    page_index: u64,
+    modified: bool,
+    data: [u8; PAGE_SIZE as usize],
+}
+
+impl StableMemoryPage {
+    fn new() -> StableMemoryPage {
+        StableMemoryPage {
+            page_index: FREE_PAGE_INDEX,
+            modified: false,
+            data: [0u8; PAGE_SIZE as usize],
+        }
+    }
+
+    fn is_free(&self) -> bool {
+        self.page_index == FREE_PAGE_INDEX
+    }
+
+    fn cached_data(&mut self) -> *mut u8 {
+        &mut self.data as *mut u8
+    }
+
+    fn stable_memory_offset(&self) -> u64 {
+        self.page_index * PAGE_SIZE
+    }
+
+    fn swap_out(&mut self) {
+        debug_assert!(!self.is_free());
+        if self.modified {
+            unsafe {
+                ic0_stable64_write(
+                    self.stable_memory_offset(),
+                    self.cached_data() as u64,
+                    PAGE_SIZE,
+                );
+            }
+        }
+        self.page_index = FREE_PAGE_INDEX;
+        self.modified = false;
+    }
+
+    fn swap_in(&mut self, page_index: u64) {
+        debug_assert!(self.is_free());
+        Self::grow_if_needed(page_index);
+        self.page_index = page_index;
+        debug_assert!(!self.modified);
+        unsafe {
+            ic0_stable64_read(
+                self.cached_data() as u64,
+                self.stable_memory_offset(),
+                PAGE_SIZE,
+            )
+        }
+    }
+
+    fn grow_if_needed(page_index: u64) {
+        grant_stable_space((page_index + 1) * PAGE_SIZE);
+    }
+}
+
+// At most two pages are needed to cache the current streamed `scan`` and `free` position.
+const CACHE_SIZE: usize = 2;
+
+/// Buffered streamed reader/writer on stable memory.
+/// Optimized for Cheney's algorithm.
+///
+/// The buffer supports two location-independent streams:
+/// * Streamed reading and updating, used for scanning and patching pointers.
+/// * Streamed writing, used for allocating new objects.
+pub struct StableMemorySpace {
+    /// The pointers used in the serialized stable memory layout are
+    /// relative to this start address of the to-space.
+    base_address: u64,
+    /// Used fore reading and updating.
+    scan_address: u64,
+    /// Used for writing.
+    free_address: u64,
+    // TODO: Use a way to not allocate a 128K empty data segment.
+    cache: [StableMemoryPage; CACHE_SIZE],
```

Comment:
Could you just use a global array? But we have 2MB of shadow stack, so 128K not so bad, unless you are worried about the initialization cost. Maybe you could just use physical page 0 and 1, and (pray) the stack doesn't grow into it.

Reply:
I was primarily worried about a potential Wasm data segment being created by the Rust compiler to define the zeroed memory map. I will check how the compiler behaves, and/or better use a more explicit initialization (not to be surprised in the future by Rust compiler changes).

Reply:
**Update**: I removed the stable memory cache, as it seems to be less efficient than direct accesses.